crates/oci/src/loader.rs

```diff
@@ -34,42 +46,58 @@ impl OciLoader {
             .lockfile_path(&reference)
             .await
             .context("cannot get path to spin.lock")?;
-        self.load_from_cache(lockfile_path, reference, &client.cache)
+        self.load_from_cache(manifest, lockfile_path, reference, &client.cache)
             .await
     }
 
     /// Loads an OCI Artifact from the given cache and returns a LockedApp with the given reference
     pub async fn load_from_cache(
         &self,
+        manifest: oci_distribution::manifest::OciImageManifest,
         lockfile_path: PathBuf,
         reference: &str,
         cache: &Cache,
-    ) -> std::result::Result<LockedApp, anyhow::Error> {
+    ) -> std::result::Result<ExecutableArtefact, anyhow::Error> {
         let locked_content = tokio::fs::read(&lockfile_path)
             .await
             .with_context(|| format!("failed to read from {}", quoted_path(&lockfile_path)))?;
-        let mut locked_app = LockedApp::from_json(&locked_content).with_context(|| {
-            format!(
-                "failed to decode locked app from {}",
-                quoted_path(&lockfile_path)
-            )
-        })?;
-
-        // Update origin metadata
-        let resolved_reference = Reference::try_from(reference).context("invalid reference")?;
-        let origin_uri = format!("{ORIGIN_URL_SCHEME}:{resolved_reference}");
-        locked_app
-            .metadata
-            .insert("origin".to_string(), origin_uri.into());
-
-        for component in &mut locked_app.components {
-            self.resolve_component_content_refs(component, cache)
-                .await
-                .with_context(|| {
-                    format!("failed to resolve content for component {:?}", component.id)
-                })?;
+        let locked_json: serde_json::Value = serde_json::from_slice(&locked_content)
+            .with_context(|| format!("OCI config {} is not JSON", quoted_path(&lockfile_path)))?;
+
+        if locked_json.get("spin_lock_version").is_some() {
+            let mut locked_app = LockedApp::from_json(&locked_content).with_context(|| {
```

Comment:
This chunk (to line 89) has not changed from line 51ff in the existing code: it has just moved down and been indented because of going inside the "is it a LockedApp" check.  If I've made this function too long and/or intricate (as I strongly suspect I have), yell at me and I'll split the arms of the `if` into separate helpers.